src/shogun/lib/auto_initialiser.h

```diff
@@ -7,61 +7,121 @@
 			~GammaFeatureNumberInit() override = default;
 
-			Any operator()() override
+			Any operator()() const final
 			{
-				require(m_kernel != nullptr, "m_kernel is not pointing to a Kernel object");
+				auto lhs = m_kernel.get_lhs();
+				switch (lhs->get_feature_class())
+				{
+				case EFeatureClass::C_DENSE:
+				case EFeatureClass::C_SPARSE:
+				{
+					auto dot_features = lhs->as<DotFeatures>();
+					return make_any(
+					    1.0 / (static_cast<float64_t>(
+					               dot_features->get_dim_feature_space()) *
+					           dot_features->get_std(false)[0]));
+				}
+				default:
+				{
+					auto dot_features = lhs->as<DotFeatures>();
+					return make_any(
+					    1.0 / static_cast<float64_t>(
+					              dot_features->get_dim_feature_space()));
+				}
+				}
+			}
+
+		private:
+			Kernel& m_kernel;
+			SG_DELETE_COPY_AND_ASSIGN(GammaFeatureNumberInit);
+		};
+
+		class GaussianWidthAutoInit : public AutoInit
+		{
+			static constexpr std::string_view kName = "GaussianWidthInit";
+			static constexpr std::string_view kDescription =
+			    "Automatic initialisation of the kernel log width "
+			    "using the median of the pairwise euclidean distance "
+			    "of all the features.";
+
+		public:
+			explicit GaussianWidthAutoInit(GaussianKernel& kernel)
+			    : AutoInit(kName, kDescription), m_kernel(kernel)
+			{
+			}
+
+			~GaussianWidthAutoInit() override = default;
+
+			Any operator()() const final
+			{
+				auto lhs = m_kernel.get_lhs();
+				auto rhs = m_kernel.get_rhs();
 
-				auto lhs = m_kernel->get_lhs();
 				switch (lhs->get_feature_class())
 				{
-					case EFeatureClass::C_DENSE:
-					case EFeatureClass::C_SPARSE:
-					{
-						auto dot_features = lhs->as<DotFeatures>();
-						return make_any(
-						    1.0 /
-						    (static_cast<float64_t>(
-								dot_features->get_dim_feature_space()) *
-								dot_features->get_std(false)[0]));
-					}
-					default:
+				case EFeatureClass::C_DENSE:
+				{
+					auto result = SGVector<float64_t>(
+					    (lhs->get_num_vectors() * lhs->get_num_vectors() -
+					     lhs->get_num_vectors()) /
+					    2);
+
+					// copy upper triangular wihout a particular order
+					index_t idx = 0;
+					for (int i = 0; i < rhs->get_num_vectors(); ++i)
 					{
-						auto dot_features = lhs->as<DotFeatures>();
-						return make_any(
-							1.0 /
-							static_cast<float64_t>(
-								dot_features->get_dim_feature_space()));
+						for (int j = i + 1; j < lhs->get_num_vectors(); ++j)
+						{
+							result[idx] = m_kernel.distance(i, j);
+							++idx;
+						}
 					}
+					return make_any(
+					    GaussianKernel::to_log_width(linalg::median(result)));
+				}
+				default:
+				{
+					float64_t default_value =
+					    GaussianKernel::to_log_width(m_kernel.get_width());
+					return make_any(default_value);
+				}
 				}
 			}
 
 		private:
-			Kernel* m_kernel;
-			SG_DELETE_COPY_AND_ASSIGN(GammaFeatureNumberInit);
+			GaussianKernel& m_kernel;
```

Comment:
I made this change so we can avoid a check for nullptr (a reference can never be null so we know that what we passed to the constructor has to be a valid object). Ideally this would be `const` but get_lhs/rhs are not const